### PR TITLE
Feature/let user leave group

### LINKED ISSUE #367: Let user leave group
## Motivations
Let the user leave groups he is member of

## Subtasks
In "My groups" page, "The teams you joined" table:
- rename "The teams you joined" to "The groups you joined"
- filter out (don't display) the group with type "Base"
- add a column on the right with a "fa sign-out-alt" icon  
- on click, display a primeng ConfirmPopup "Are you sure you want to leave this group?"
  * call [groupLeave](https://france-ioi.github.io/algorea-devdoc/api/#operation/groupLeave) on yes
    *  display a green/red toast message (same as the other services) for indicating success/failure
- once done, refresh the table (use the `resetter` option of `mapToFetchState`)

### PR DESCRIPTION
**Pull request by Iloveall**
Fixes #367

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
Initial UI-testing feedback:
- There are sort icons are on the "action" column while only "joined on" is sortable. (and clicking on the right of the action column apply sorting)
Screenshot 2021-04-29 at 09 33 33
- This is not a [primeng ConfirmPopup](https://www.primefaces.org/primeng/showcase/#/confirmpopup) as requested
- Of course, the group name should be indicated, not it's ID or whatever. (and the text should be `You have left "group name"`
Screenshot 2021-04-29 at 09 37 38

**Comment by smadbe** [comment]
Could you put quote around the group name ? So `You have left "123"`
Screenshot 2021-04-29 at 12 21 13

File: src/app/core/http-services/joined-groups.service.ts [hunk 1]
```diff
@@ -46,6 +52,8 @@ export class JoinedGroupsService {
       );
   }
 
-}
-
+  leave(id: string): Observable<LeaveGroupResponse> {
```
Comment (smadbe) [inline src/app/core/http-services/joined-groups.service.ts:55]: To make thing more clear:
```
  leave(groupId: string): Observable<LeaveGroupResponse> {
```

File: src/app/core/http-services/joined-groups.service.ts [hunk 2]
```diff
@@ -46,6 +52,8 @@ export class JoinedGroupsService {
...
+    return this.http.delete<LeaveGroupResponse>(`${appConfig().apiUrl}/current-user/group-memberships/${id}`);
```
Comment (smadbe) [inline src/app/core/http-services/joined-groups.service.ts:56]: For POST/PUT/DELETE, you have to use to map to `successData` instead... and here, probably map to just a boolean. Search for usage of `successData` at other places in the project...

File: src/app/core/http-services/joined-groups.service.ts [hunk 3]
```diff
@@ -46,6 +52,8 @@ export class JoinedGroupsService {
...
+  }
 
+}
```
Comment (smadbe) [inline src/app/core/http-services/joined-groups.service.ts:59]: This file was specifically for the service returning joined group. So if you cannot just add this service here.
Probably it is easier to just create a "GroupLeaveService".

**Review by smadbe — CHANGES_REQUESTED** [review]
I haven't completed this review (I have to go) but here are already my first remarks

File: src/app/core/http-services/joined-groups.service.ts (hunk 1 above)
Comment (Iloveall) [reply src/app/core/http-services/joined-groups.service.ts]: Fixed

File: src/app/core/http-services/joined-groups.service.ts (hunk 2 above)
Comment (Iloveall) [reply src/app/core/http-services/joined-groups.service.ts]: Fixed

File: src/app/core/http-services/joined-groups.service.ts (hunk 3 above)
Comment (Iloveall) [reply src/app/core/http-services/joined-groups.service.ts:49]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts [hunk 4]
```diff
@@ -55,6 +55,7 @@ export class JoinedGroupListComponent implements OnDestroy {
     const groupId = group.group.id;
     const groupName = group.group.name;
     this.joinedGroupsService.leave(groupId)
+      .pipe(first())
```
Comment (smadbe) [inline src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:60]: I don't get the point of this. All http services only emit once and then complete.

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts [hunk 5]
```diff
@@ -32,4 +38,42 @@ export class JoinedGroupListComponent implements OnDestroy {
...
+      .subscribe(() => {
+        this.refresh$.next();
+        this.messageService.add({
+          severity: 'success',
+          summary: $localize`Success`,
+          detail: $localize`You have left ${groupName}`,
+          life: TOAST_LENGTH,
+        });
+      }, _err => {
```
Comment (smadbe) [inline src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:69]: It looks like there is a wrong indent here.
Should be:
```
.subscribe(
  () => {
    ...
  },
  _err => {
    ...
  }
)
```

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts [hunk 6]
```diff
@@ -32,4 +38,42 @@ export class JoinedGroupListComponent implements OnDestroy {
...
+    this.groupLeaveService.leave(groupId)
+      .pipe(first())
+      .subscribe(() => {
+        this.refresh$.next();
+        this.messageService.add({
+          severity: 'success',
+          summary: $localize`Success`,
+          detail: $localize`You have left ${groupName}`,
+          life: TOAST_LENGTH,
+        });
```
Comment (smadbe) [inline src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:68]: you'll have to get master (`merge master`) and use the new service.

**Review by smadbe — CHANGES_REQUESTED** [review]
👍  Only a few minor changes.

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts (hunk 4 above)
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:60]: Yes, l got TS notification from IDE but right now it's gone, strange... I'll remove then

File: src/app/core/http-services/group-leave.service.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { appConfig } from 'src/app/shared/helpers/config';
+import { ActionResponse, successData } from 'src/app/shared/http-services/action-response';
+
+interface LeaveGroupResponse {
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/core/http-services/group-leave.service.ts:8]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts [hunk 8]
```diff
@@ -1,27 +1,33 @@
...
+import { GroupLeaveService } from 'src/app/core/http-services/group-leave.service';
 
 @Component({
   selector: 'alg-joined-group-list',
   templateUrl: './joined-group-list.component.html',
   styleUrls: [ './joined-group-list.component.scss' ]
 })
 export class JoinedGroupListComponent implements OnDestroy {
-
+  refresh$ = new Subject();
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:17]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts [hunk 9]
```diff
@@ -1,27 +1,33 @@
...
   private readonly sort$ = new ReplaySubject<SortOptions>(1);
   readonly state$ = this.sort$.pipe(
     startWith(NO_SORT),
     distinctUntilChanged(sortEquals),
     switchMap(sort => this.joinedGroupsService.getJoinedGroups(sort)),
-    mapToFetchState(),
+    map(group => group.filter(g => g.group.type !== 'Base')),
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:23]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts (hunk 4 above)
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:60]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts (hunk 5 above)
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts:69]: Fixed

File: src/app/modules/group/components/joined-group-list/joined-group-list.component.ts (hunk 6 above)
Comment (Iloveall) [reply src/app/modules/group/components/joined-group-list/joined-group-list.component.ts]: Fixed

**Review by smadbe — APPROVED** [review]

**Pull request merged**